src/6model/reprs/VMArray.c

```diff
@@ -792,15 +792,99 @@ static void shift(MVMThreadContext *tc, MVMSTable *st, MVMObject *root, void *da
+                                    ? (MVMArrayREPRData *)STABLE(dest)->REPR_data : NULL;
+
+    if (elems > 0) {
+        MVMint64  i;
+        MVMuint16 kind;
+        MVMuint8 d_needs_barrier = dest->header.flags & MVM_CF_SECOND_GEN;
+        if (s_repr_data && d_repr_data
+                && s_repr_data->slot_type == d_repr_data->slot_type
+                && s_repr_data->elem_size == d_repr_data->elem_size
+                && (d_repr_data->slot_type != MVM_ARRAY_OBJ || !d_needs_barrier)
+                && d_repr_data->slot_type  != MVM_ARRAY_STR) {
+            /* Optimized for copying from a VMArray with same slot type */
+            MVMint64 s_start = s_body->start;
+            MVMint64 d_start = d_body->start;
+            memcpy( d_body->slots.u8 + (d_start + d_offset) * d_repr_data->elem_size,
+                    s_body->slots.u8  + (s_start + s_offset) * s_repr_data->elem_size,
+                    d_repr_data->elem_size * elems
+            );
+        }
+        else {
+            switch (s_repr_data->slot_type) {
+                case MVM_ARRAY_OBJ:
+                    kind = MVM_reg_obj;
+                    break;
+                case MVM_ARRAY_STR:
+                    kind = MVM_reg_str;
+                    break;
+                case MVM_ARRAY_I64:
+                case MVM_ARRAY_I32:
+                case MVM_ARRAY_I16:
+                case MVM_ARRAY_I8:
+                    kind = MVM_reg_int64;
+                    break;
+                case MVM_ARRAY_N64:
+                case MVM_ARRAY_N32:
+                    kind = MVM_reg_num64;
+                    break;
+                case MVM_ARRAY_U64:
+                case MVM_ARRAY_U32:
+                case MVM_ARRAY_U16:
+                case MVM_ARRAY_U8:
+                    kind = MVM_reg_int64;
+                    break;
+                default:
+                    abort(); /* never reached, silence compiler warnings */
+            }
+            for (i = 0; i < elems; i++) {
+                MVMRegister to_copy;
+                REPR(src)->pos_funcs.at_pos(tc, STABLE(src), src, s_body, s_offset + i, &to_copy, kind);
+                bind_pos(tc, STABLE(dest), dest, d_body, d_offset + i, to_copy, kind);
+            }
+        }
+    }
+}
+
+static void aslice(MVMThreadContext *tc, MVMSTable *st, MVMObject *src, void *data, MVMObject *dest, MVMint64 start, MVMint64 end) {
+    MVMArrayBody     *s_body      = (MVMArrayBody *)data;
+    MVMArrayBody     *d_body      = (MVMArrayBody *)OBJECT_BODY(dest);
+    MVMArrayREPRData *d_repr_data = REPR(dest)->ID == MVM_REPR_ID_VMArray
+                                      ? STABLE(dest)->REPR_data : NULL;
+
+    MVMint64 last_elem = REPR(src)->elems(tc, st, src, s_body) - 1;
+    MVMint64 elems;
+
+    if ( last_elem < start || last_elem < end
+         || (end < start && 0 <= start && 0 <= end) )
+    {
+        MVM_exception_throw_adhoc(tc, "MVMArray: Slice Index Out of Bounds");
```

Comment:
I would lowercase "Index Out of Bounds" to be more consistent with the rest of the "MVMArray:" errors.